web/frontend/src/components/slam.vue

```diff
@@ -1,193 +1,221 @@
-let slam2dIntervalId = -1;
+let show3d = $ref(false);
+let refresh2DCancelled = true;
+let refresh3DCancelled = true;
 
-/*
- * const selected3dValue = $ref('manual');
- * let pointCloudUpdateCount = $ref(0);
- * let pointcloud = $ref<Uint8Array | undefined>();
- * let pose = $ref<commonApi.Pose | undefined>();
- * let show3d = $ref(false);
- * const loaded2d = $computed(() => (pointcloud !== undefined && pose !== undefined));
- * let slam3dIntervalId = -1;
- */
+const loaded2d = $computed(() => (pointcloud !== undefined && pose !== undefined));
 
-/*
- * const concatArrayU8 = (arrays: Uint8Array[]) => {
- *   const totalLength = arrays.reduce((acc, value) => acc + value.length, 0);
- *   const result = new Uint8Array(totalLength);
- *   let length = 0;
- *   for (const array of arrays) {
- *     result.set(array, length);
- *     length += array.length;
- *   }
- *   return result;
- * };
- */
+let slam2dTimeoutId = -1;
+let slam3dTimeoutId = -1;
 
-const fetchSLAMImageMap = (name: string) => {
-  const req = new slamApi.GetMapRequest();
-  req.setName(name);
-  req.setMimeType('image/jpeg');
-  req.setIncludeRobotMarker(true);
-  rcLogConditionally(req);
-  props.client.slamService.getMap(req, new grpc.Metadata(), (error, response) => {
-    if (error) {
-      return displayError(error);
-    }
-    const blob = new Blob([response!.getImage_asU8()], { type: 'image/jpeg' });
-    imageMap = URL.createObjectURL(blob);
-  });
+const concatArrayU8 = (arrays: Uint8Array[]) => {
+  const totalLength = arrays.reduce((acc, value) => acc + value.length, 0);
+  const result = new Uint8Array(totalLength);
+  let length = 0;
+  for (const array of arrays) {
+    result.set(array, length);
+    length += array.length;
+  }
+  return result;
 };
 
-/*
- * const fetchSLAMMap = (name: string) => {
- *   return nextTick(() => {
- *     const req = new slamApi.GetPointCloudMapStreamRequest();
- *     req.setName(name);
- *     rcLogConditionally(req);
- *     const chunks: Uint8Array[] = [];
- *     const getPointCloudMapStream = props.client.slamService.getPointCloudMapStream(req);
- *     getPointCloudMapStream.on('data', (res) => {
- *       const chunk = res.getPointCloudPcdChunk_asU8();
- *       chunks.push(chunk);
- *     });
- *     getPointCloudMapStream.on('status', (status) => {
- *       if (status.code !== 0) {
- *         const error = {
- *           message: status.details,
- *           code: status.code,
- *           metadata: status.metadata,
- *         };
- *         displayError(error);
- *       }
- *     });
- *     getPointCloudMapStream.on('end', (end) => {
- *       if (end === undefined || end.code !== 0) {
- *         // the error will be logged in the 'status' callback
- *         return;
- *       }
- *       pointcloud = concatArrayU8(chunks);
- *       pointCloudUpdateCount += 1;
- *     });
- *   });
- * };
- */
+const fetchSLAMMap = (name: string): Promise<Uint8Array> => {
+  return new Promise((resolve, reject) => {
```

Comment:
Why does this need to be async? Since we always immediately call `await` on the result, I think we're not using the async property of it.

Reply:
@micheal-parks correct me if I'm wrong, but I think this needs to be async so that we don't block the event loop while waiting for the responses from the grpc callback functions fetchSLAMMap & fetchSLAMPose.

Reply:
But won't we block the event loop during the `await` in `updateSLAM2dRefreshFrequency`?

Reply:
Neither will block the event loop. The await will allow you to create imperative logic step by step. Any time net IO happens in JS, it will yield to the event loop. Once the IO is done, a callback or promise will be resolved and scheduled onto the event loop.

Really the only time you can block is if you have tight compute logic (like a while true).

Reply:
What do you mean by "the await will allow you to create imperative logic step by step"? I don't have much UI experience, but my experience with futures/promises in C++ is that there's no reason to make something async if callers always immediately block on the result.

Reply:
You can't block on a call in Javascript. Fetch/XMLHttpRequest is always asynchronous. Under the covers, `await`ing anything that is a `Promise` will ask the JS runtime to return to the line after the await once the underlying promise is `resolve`d or `reject`ed.

What I mean by imperative is that when using the async/await pattern in a language like JavaScript, you can read the code as if it was blocking, step by step, but it actually lets the runtime and event loop do its work. Everything is running on 1 thread in a your default JavaScript event loop so the language itself is made to be as asynchronous/callback based as possible. The only time you'd see truly blocking behavior is if an API says so and it really only happens when running in a server context (like in node).

Reply:
Keep in mind there's a ton of other code running here making other requests. The await lets one programmer "block" without blocking others from completing their work.

Reply:
Funny story, you can make XMLHttpRequest synchronous, but it's terribly bad practice unless done in a web worker. I've seen it done before 😰  If the request is made on the main thread - which in JS is the rendering thread - it'll freeze the entire tab until its completed.